toolchain/driver/clang_runner.cpp

```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+                                  resource_dir_path / "share", ec);
+
+  std::filesystem::path lib_path = resource_dir_path / "lib";
+  std::filesystem::create_directory(lib_path, ec);
+  lib_path /= target.str();
+  std::filesystem::create_directory(lib_path, ec);
+
+  llvm::Triple target_triple(target);
+  CARBON_CHECK(!target_triple.isOSWindows(),
+               "TODO: Windows runtimes are untested and not yet supported.");
+
+  // For Linux targets, the system libc (typically glibc) doesn't necessarily
+  // provide the CRT begin/end files, and so we need to build them.
+  if (target_triple.isOSLinux()) {
+    std::filesystem::path begin_o_path = lib_path / "clang_rt.crtbegin.o";
+    BuildCrtFile(target, RuntimeSources::CrtBegin, begin_o_path.native());
+
+    std::filesystem::path end_o_path = lib_path / "clang_rt.crtend.o";
+    BuildCrtFile(target, RuntimeSources::CrtEnd, end_o_path.native());
+  }
+
+  llvm::SmallVector<llvm::StringRef, 512> src_files;
+  auto append_src_files = [&src_files](auto input_srcs) {
+    for (llvm::StringRef input_src : input_srcs) {
+      if (input_src.ends_with(".c") || input_src.ends_with(".S")) {
+        src_files.push_back(input_src);
+      }
+    }
+  };
+  append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsGenericSrcs));
+  append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsBf16Srcs));
+  if (target_triple.isArch64Bit()) {
+    append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsTfSrcs));
+  }
+  if (target_triple.isAArch64()) {
+    append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsAarch64Srcs));
+  } else if (target_triple.isX86()) {
+    append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsX86ArchSrcs));
+    if (target_triple.isArch64Bit()) {
+      append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsX86_64Srcs));
+    } else {
+      CARBON_CHECK(
+          target_triple.isArch32Bit(),
+          "The Carbon toolchain doesn't currently support 16-bit x86.");
+      append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsI386Srcs));
+    }
+  } else {
+    CARBON_FATAL("Target architecture is not supported: {0}", target);
+  }
+
+  // We need a separate directory for object files used in the builtins archive.
+  std::filesystem::path objs_path = tmp_dir_ / "_objs";
+  std::filesystem::create_directory(objs_path, ec);
+
+  llvm::SmallVector<llvm::NewArchiveMember, 512> objs;
+  objs.reserve(src_files.size());
+  for (llvm::StringRef src_file : src_files) {
+    llvm::StringRef obj_file = src_file;
+    std::filesystem::path obj_path = objs_path;
+    for (;;) {
+      auto slash_offset = obj_file.find('/');
+      if (slash_offset == llvm::StringRef::npos) {
+        break;
+      }
+      obj_path.append(obj_file.begin(), obj_file.begin() + slash_offset);
+      std::filesystem::create_directory(obj_path, ec);
+      obj_file = obj_file.drop_front(slash_offset + 1);
+    }
+    obj_path.append(obj_file.begin(), obj_file.end());
```

Comment:
Is this intended to create the parent directory structure? Had you considered something like:

```suggestion
    std::filesystem::path obj_path = src_file;
    std::filesystem::create_directories(obj_path.parent_path(), ec);
```

Reply:
I seem to have forgotten about `create_directories` entirely and re-implemented.. doh!